examples/apps/ai-collab/src/components/TaskGroup.tsx

```diff
@@ -59,8 +68,91 @@ export function TaskGroup(props: {
 		newBranchTargetNode: SharedTreeTaskGroup;
 	}>();
 
+	const [undoStack, setUndoStack] = useState<Revertible[]>([]);
+	const [redoStack, setRedoStack] = useState<Revertible[]>([]);
+
 	useSharedTreeRerender({ sharedTreeNode: props.sharedTreeTaskGroup, logId: "TaskGroup" });
 
+	/**
+	 * Create undo and redo stacks of {@link Revertible}.
+	 */
+	useEffect(() => {
+		function onRevertibleDisposed(disposed: Revertible): void {
+			const redoIndex = redoStack.indexOf(disposed);
+			if (redoIndex === -1) {
+				const undoIndex = undoStack.indexOf(disposed);
+				if (undoIndex !== -1) {
+					setUndoStack((currUndoStack) => {
+						const newUndoStack = [...currUndoStack];
+						newUndoStack.splice(undoIndex, 1);
+						return newUndoStack;
+					});
+				}
+			} else {
+				setRedoStack((currRedostack) => {
+					const newRedoStack = [...currRedostack];
+					newRedoStack.splice(redoIndex, 1);
+					return newRedoStack;
+				});
+			}
+		}
+
+		/**
+		 * Instead of application developer manually managing the life cycle of the {@link Revertible} instances,
+		 * example app stores upto `MAX_STACK_SIZE` number of {@link Revertible} instances in the undo and redo stacks.
+		 * When the stack size exceeds `MAX_STACK_SIZE`, the oldest {@link Revertible} instance is disposed.
+		 * @param stack - The primary stack that the {@link Revertible} instance is being added to.
+		 * @param setstack - The setter function for the primary stack.
+		 */
+		function maintainStackSize(
+			stack: Revertible[],
+			setStack: React.Dispatch<React.SetStateAction<Revertible[]>>,
+		): void {
+			const MAX_STACK_SIZE = 5;
+
+			if (stack.length > MAX_STACK_SIZE) {
+				console.log("Stack size exceeded! Disposing oldest revertible.");
+				const oldestRevertible = stack[0];
+				if (oldestRevertible) {
+					if (oldestRevertible.status !== RevertibleStatus.Disposed) {
+						oldestRevertible.dispose();
+					}
+					// Use functional updates to ensure we're working with the latest state
+					setStack((currentStack) => {
+						const newStack = currentStack.filter((item) => item !== oldestRevertible);
+						return newStack;
+					});
+				}
+			}
+		}
+
+		const handleCommitApplied = props.treeView.events.on(
```

Comment:
Nit: this variable is actually the "off" handler for the event we registered. I'm not sure what our normal naming convention is for this pattern, but it would probably be worth renaming for clarity.

Reply:
What do you think about `treeEventHandler()`?